### ISSUE TITLE
[help] - Running equivalent queries to UI via the API

### ISSUE THREAD

**Issue by tlansec** [carbonblack/cbapi-python#266]
Hi,

**TLDR; How do you do simple queries from the UI via the API?**

In the UI, it's possible to create simple queries that don't consist of key value pairs using the "Advanced Search" functionality, e.g.:

![image](https://user-images.githubusercontent.com/5901749/90531707-4a742800-e16e-11ea-9e5b-89cdeb168f8b.png)

When trying to build queries using the CBAPI, it doesnt seem possible to simply copy/paste this type of query into the python API without raising "cbapi.errors.ApiError: Invalid query:" or another error:

```
from cbapi.psc.threathunter import CbThreatHunterAPI
from cbapi.psc.threathunter.models import Event

example_q = '"-EncodedCommand"'
cb = CbThreatHunterAPI()
event_query = cb.select(Event).where(example_q)
print(event_query.first())
```

It is possible however to run the same query by manually building the URI etc, e.g.:

```
import requests, json, time 

api_key_token = "" # REMOVED
org_key = "" # REMOVED
# example taken from documentation at: https://developer.carbonblack.com/reference/carbon-black-cloud/cb-threathunter/latest/process-search-v2/#start-an-enriched-events-search-job

post_body = {
  "criteria": {
  	"device_name": []
  },
  "query": "\"EncodedCommand\"",
  "terms": {
    "fields": [
    ],
    "rows": 100
  },
  "time_range": {
    "end": "2020-08-27T18:34:04Z",
    "start": "2020-08-01T18:34:04Z"
  }
}

target_url = "https://defense-prod05.conferdeploy.net/api/investigate/v2/orgs/{0}/enriched_events/search_jobs".format(org_key)
headers = {
    "X-Auth-Token" : api_key_token,
    "Content-Type" : "application/json",
    "accept" : "application/json"
}

r = requests.post(target_url, data=json.dumps(post_body), headers=headers)
print(r.status_code)
print(r.content)
jdata = json.loads(r.content)
job_id = jdata['job_id']
time.sleep(5) # allow job to finish 
result_url = target_url +"/{0}/results".format(job_id) # used to fetch results
r = requests.get(result_url, headers=headers)
print(r.status_code)
print(r.content)
```

Ideally, I'd like to do the above without re-writing my own client library - is there a way to do it using the cbapi library?

Cheers,
Tom

**Comment by tlansec** [comment]
The equivalent way to copy paste this type of query from the UI and use it in a script is:

```from cbapi.psc.threathunter import CbThreatHunterAPI,Process, QueryBuilder
cb = CbThreatHunterAPI()
query = QueryBuilder().where("notepad.exe")
entries = cb.select(Process).where(query).all()
for e in entries:
    print(e)
```

I was unable to find any equivalent example to this in the docs even though it seems like it should be the most obvious thing to want do (to me, anyway).

I also spotted a bug in the documentation:

https://cbapi.readthedocs.io/en/latest/threathunter-api.html

Code snippet under "Queries -> QueryBuilder" is not valid as it expects a string argument not a dict.

You can close this ticket, but I would generally recommend adding a lot more examples to the documentation as without these users have to read the description of every single method to understand how things work.

**Comment by cnotin** [comment]
Hi @tlansec! Which CB product are you using? The UI looks like "CB Defense" (now "Endpoint Standard") but it should not allow ThreatHunter query so I'm wondering...
Also, I found [this API](https://developer.carbonblack.com/reference/carbon-black-cloud/platform/latest/platform-search-api-enriched-events/):
> api/investigate/v2/orgs/{org_key}/enriched_events/facet_jobs
Which seems to allow normal queries:
> Query in lucene syntax and/or including value searches.

However, this does not seem to be implemented in this Python implementation (yet?)

**Comment by tlansec** [comment]
Hi,

The product in use was indeed Endpoint Standard - it seems like it shouldnt allow ThreatHunter queries but somehow it still does and if you inspect what's requested when you perform searches in the UI it uses the Threathunter API endpoint.

I did see the facet_jobs API but i think the one i used in the larger snippet was more appropriate for my use case:

https://developer.carbonblack.com/reference/carbon-black-cloud/cb-threathunter/latest/process-search-v2/#start-an-enriched-events-search-job

If you're finding its not working when you try to use the threathunter api, for me at least i found the differences between these to be confusing - so ensure you're using the right values

API Token consists of the key/token separated by a slash - should match [A-Z0-9]{24,}/[A-Z0-9]{7,}
Org Key is not the numerical value or org id, should match [A-Z0-9]{7,}

Hope this helps.

Cheers,
Tom

**Comment by cnotin** [comment]
Thanks for the confirmation! I also tried to use the API as captured from the UI before we had this API which allows full text queries :) Given the recent evolution of the UI to look more like ThreatHunter it isn't a surprise!

The facet_jobs is to get the facets: results bucketed by their values which are now visible in the left hand side column of the UI.

Regarding the link you posted, the endpoint is the same as [the one in the "Cloud Platform" section of the doc](https://developer.carbonblack.com/reference/carbon-black-cloud/platform/latest/platform-search-api-enriched-events/#start-an-enriched-events-search-v2) (which was linked from the deprecated "[find events](https://developer.carbonblack.com/reference/carbon-black-cloud/cb-defense/latest/rest-api/#find-events)" API from CB Defense) so it's the same in the end :)

Thank you very much for the tips!

**Comment by cnotin** [comment]
Thanks @tlansec for the token format it wasn't obvious at all!
`token=<secretkey>/<key ID>`

**Comment by cnotin** [comment]
So I think that this project needs to implement the [PSC Enriched Events API](https://developer.carbonblack.com/reference/carbon-black-cloud/platform/latest/platform-search-api-enriched-events/). I hope someone from CB can comment :)

**Comment by cnotin** [comment]
Hey @tlansec just wanted to share something I discovered. It looks like (#278) there is a [new CB library](https://github.com/carbonblack/carbon-black-cloud-sdk-python/) and [Enriched Events Search is being implemented](https://github.com/carbonblack/carbon-black-cloud-sdk-python/pull/60)
That might explains why it did not get implemented here!

**Comment by tlansec** [comment]
Thanks for the heads up!

**Comment by avanbrunt-cb** [comment]
Closing issue as it appears to be resolved in https://github.com/carbonblack/carbon-black-cloud-sdk-python. Please reopen if there is still an issue
